src/Psalm/Internal/Analyzer/StatementsAnalyzer.php

```diff
@@ -1464,4 +1465,36 @@ public function getFunctionAnalyzers()
     {
         return $this->function_analyzers;
     }
+
+    /**
+     * @param  PhpParser\Node\Expr $first_arg_value
+     *
+     * @return null|string
+     */
+    public static function getConstName($first_arg_value)
+    {
+        $const_name = null;
+
+        if ($first_arg_value instanceof PhpParser\Node\Scalar\String_) {
+            $const_name = $first_arg_value->value;
+        } elseif (isset($first_arg_value->inferredType)) {
+            $possible_value_types = $first_arg_value->inferredType->getTypes();
+            if (count($possible_value_types) === 1) {
+                reset($possible_value_types);
+                $possible_value_type = current($possible_value_types);
+                if ($possible_value_type instanceof Type\Atomic\TLiteralString) {
+                    $const_name = $possible_value_type->value;
+                }
+            }
+        }
+
+        if ($const_name !== null) {
+            $namespace_pos = strrpos($const_name, '\\');
```

Comment:
Why is the namespace being removed? This looks like this would attempt to add `SomeConst` instead of `Ns\SomeConst` for `define('Ns\SomeConst', 'value')` (from any namespace)

Reply:
Adding `setFullyQualifiedConstType($fully_qualified_const_name, Type\Union $const_type)` would probably be the simplest fix in src/Psalm/Internal/Analyzer/NamespaceAnalyzer.php and StatementsAnalyzer.php (or rethink $context->constants)
 
```
    /**
     * @param string     $const_name
     * @param Type\Union $const_type
     *
     * @return void
     */
    public function setConstType($const_name, Type\Union $const_type)
    {
        self::$public_namespace_constants[$this->namespace_name][$const_name] = $const_type;
    }
```

Reply:
Hmm. I didn't realize that two of the files call that function, meaning they do use a namespace already. Wouldn't it be the correct namespace, meaning that `setFullyQualifiedConstType()` is unnecessary? The only file that doesn't use that method is src/Psalm/Internal/Visitor/ReflectorVisitor.php, which is used the most.

I'm not sure why, but `define()` always ignores namespaces from user code, so this is nothing new. I do agree that this should be fixed, but I'm not sure if it makes sense for it to happen in this PR. This fixes a lot of issues, and I'm pretty sure that fixing `define()` to consider namespaces would be more involved.